module/zfs/mmp.c

```diff
@@ -451,6 +452,19 @@ mmp_write_uberblock(spa_t *spa)
 		    "gethrtime %llu", spa_name(spa), lock_acquire_time,
 		    gethrtime());
 
+	/*
+	 * Skip mmp uberblock writing if raidz expansion is in progress and
+	 * scratch object is active.
```

Comment:
Why would we want to skip the mmp write?  Are you thinking that the mmp uberblock would be overwriting the scratch object?  That should not be the case; the mmp uberblock is inside the uberblock ring, i.e. `vdev_label_t:vl_uberblock`, which is just before the scratch/boot region.  See `VDEV_BOOT_OFFSET`, and `VDEV_UBERBLOCK_OFFSET()`.

Reply:
It does not overwrite scratch object, I've checked it. But the blocks corruptions have place, if
mmp writing is not disabled. I think that mmp uberblock produce some sort of scratch value inconsistency,
when during single expansion scratch object is enabled, then disabled by mmp uberblock writing, and then enabled again
by importing using regular uberblock. This situation cause bp corruptions. The another workaround is to not
actualize mmp ub_timestamp in the mmp code. From other side the same ub_timestamp update is doing for uberblocks under scratch code.
I've updated the comment little bit to make it more clean.

Reply:
I think the problem is that the raidz code overwrites the uberblock with the SCRATCH_VALID bit set, but does not notify the mmp code.  Then when the MMP code writes a mmp uberblock, it doesn't have the SCRATCH_VALID bit set.  If we crash, we'll use the MMP uberblock since it was written more recently.

Disabling MMP uberblock updates would mean that we could lose our lease on the pool, and another system could import it, which would be bad.

I think the better solution would be to notify the MMP code when the raidz code updates the uberblock on disk, so that MMP writes will take this in to account.  i.e. after calling `vdev_uberblock_sync_list()`, we need to call `mmp_update_uberblock()`, like `vdev_config_sync()` does.  This would need to happen in `raidz_reflow_scratch_sync()` and `vdev_raidz_reflow_copy_scratch()`.

Reply:
Done, as you suggested.